tt_metal/impl/device/device.cpp

```diff
@@ -3229,6 +3229,15 @@ void Device::set_worker_mode(const WorkExecutorMode& mode) {
 void Device::enable_async(bool enable) {
     auto mode = enable ? WorkExecutorMode::ASYNCHRONOUS : WorkExecutorMode::SYNCHRONOUS;
     this->set_worker_mode(mode);
+    // If a worker thread is spawned for a device, register/track it in a runtime structure.
```

Comment:
Is the contract that you can only call this from the main thread?

Reply:
Hey @ayerofieiev-tt, the contract is that this function cannot be called from worker threads that Metal backend spawns. A user can choose to do the following:

```
1. Construct a device object in the main thread
2. Fork an application thread off the main thread
3. Call enable_async on the device in the forked thread
4. Run a workload using the device in the forked thread
```

Or

```
1. Construct a device object in the main thread and call enable_async on the device in the main thread
2. Fork an application thread and run a workload in the forked thread
```

The latter is what the models team does for integrating Metal with VLLM, but either case is supported.

Reply:
> this function cannot be called from worker threads that Metal backend spawns

Please consider leaving this as a comment to this function in .hpp, fairly important note

Reply:
Future tech report in the making to explain the contract and enumerate cases for async 😄